Project-Hotpot/Project-Hotpot/Managers/CacheManager.swift

```diff
@@ -0,0 +1,128 @@
+                      let items = tracks?["items"] as? [[String:Any]?]
+                else{
+                    NSLog("failed parsing random song dictionary response")
+                    return completion(.failure(CustomError.failedResponseParsing))
+                }
+                let trackIDs : [String] = items.compactMap{ item in item?["id"] as? String }
+                self.spotifyIdToSongDetails(ids: trackIDs) { result in
+                    switch result{
+                    case .success(let songDetailsArray):
+                        self.cache.addToCache(genre: genre, songs: songDetailsArray)
+                        completion(.success(songDetailsArray))
+                    case .failure(let error) :
+                        completion(.failure(error))
+                    }
+                }
+            }
+        }
+    }
+    
+    func spotifyIdToSongDetails(ids: [String], completion: @escaping (_ result: Result<[SongDetails], Error>) -> Void) {
+        SpotifyManager.shared().fetchAudioFeaturesFromTracks(for: ids) { result in
+            switch result{
+            case .success(let dictionary):
+                guard let features = dictionary["audio_features"] as? [[String:Any]]
+                else {
+                    NSLog("failed to extract features dictionary")
+                    return
+                    
+                }
+                let songDetailsArray:[SongDetails]? = features.compactMap { feature in
+                    guard let featureDictionary = feature as? [String: Any],
+                          let uri = featureDictionary["uri"] as? String,
+                          let id = featureDictionary["id"] as? String,
+                          let danceability = featureDictionary["danceability"] as? NSNumber,
+                          let energy = featureDictionary["energy"] as? NSNumber,
+                          let tempo = featureDictionary["tempo"] as? NSNumber,
+                          let key = featureDictionary["key"] as? NSNumber
+                    else {
+                        NSLog("failed parse of audio feature dictionary")
+                        return nil
+                    }
+                    return SongDetails(uri: uri, id: id, danceability: danceability, energy: energy, tempo: tempo, key: key)
+                }
+                if let songDetailsArray = songDetailsArray{
+                    return completion(.success(songDetailsArray))
+                }
+                else {
+                    return completion(.failure(CustomError.failedResponseParsing))
+                }
+                
+            case .failure(let error):
+                NSLog("failed audio features: \(error)")
+                completion(.failure(error))
+            }
+        }
+    }
+    
+    func fetchNSongs (limit: Int, genre: String, completion: @escaping (_ result: Result<[String:Any], Error>) -> Void) {
+        let randomOffset = Int.random(in: 1..<800)
```

Comment:
Consider adding a seed to make it more non deterministic (e.q. current timestamp)